src/main/java/ch/njol/skript/expressions/ExprElement.java

```diff
@@ -142,36 +204,50 @@ public Class<?> getReturnType() {
 	}
 
 	@Override
-	public String toString(@Nullable Event e, boolean debug) {
-		String prefix;
+	public String toString(@Nullable Event event, boolean debug) {
+		String prefix = "";
 		switch (type) {
-			case FIRST:
+			case FIRST_ELEMENT:
 				prefix = "the first";
 				break;
-			case LAST:
+			case LAST_ELEMENT:
 				prefix = "the last";
 				break;
+			case FIRST_X_ELEMENTS:
+				assert startIndex != null;
+				prefix = "the first " + startIndex.toString(event, debug);
+				break;
+			case LAST_X_ELEMENTS:
+				assert startIndex != null;
+				prefix = "the last " + startIndex.toString(event, debug);
+				break;
 			case RANDOM:
 				prefix = "a random";
 				break;
 			case ORDINAL:
-				assert number != null;
+			case TAIL_END_ORDINAL:
+				assert startIndex != null;
 				prefix = "the ";
 				// Proper ordinal number
-				if (number instanceof Literal) {
-					Number number = ((Literal<Number>) this.number).getSingle();
+				if (startIndex instanceof Literal) {
+					Number number = ((Literal<Number>) startIndex).getSingle();
 					if (number == null)
-						prefix += this.number.toString(e, debug) + "th";
+						prefix += startIndex.toString(event, debug) + "th";
 					else
 						prefix += StringUtils.fancyOrderNumber(number.intValue());
 				} else {
-					prefix += number.toString(e, debug) + "th";
+					prefix += startIndex.toString(event, debug) + "th";
 				}
+				if (type == ElementType.TAIL_END_ORDINAL)
+					prefix += " last";
 				break;
+			case RANGE:
+				assert startIndex != null && endIndex != null;
+				return "the elements from " + startIndex.toString(event, debug) + " to " + endIndex.toString(event, debug) + " of " + expr.toString(event, debug);
 			default:
 				throw new IllegalStateException();
```

Comment:
Just remove this default, it makes no sense. I just encountered an issue unrelated to this pull request with the existing implementation of this class. I came to see if this pull request had fixed it, and it seems it has, but this default throwing an exception makes no sense when the enum is local, and all types are implemented.

Reply:
```
[02:43:38 ERROR]: #!#! Stack trace:
[02:43:38 ERROR]: #!#! java.lang.IllegalStateException
[02:43:38 ERROR]: #!#!     at Skript.jar//ch.njol.skript.expressions.ExprElement.toString(ExprElement.java:172)
[02:43:38 ERROR]: #!#!     at Skript.jar//ch.njol.skript.conditions.CondCompare.toString(CondCompare.java:336)
[02:43:38 ERROR]: #!#!     at Skript.jar//ch.njol.skript.test.runner.EffAssert.toString(EffAssert.java:105)
[02:43:38 ERROR]: #!#!     at Skript.jar//ch.njol.skript.ScriptLoader.loadItems(ScriptLoader.java:939)
[02:43:38 ERROR]: #!#!     at Skript.jar//ch.njol.skript.lang.SkriptEvent.load(SkriptEvent.java:133)
[02:43:38 ERROR]: #!#!     at Skript.jar//ch.njol.skript.ScriptLoader.lambda$loadScripts$8(ScriptLoader.java:557)
[02:43:38 ERROR]: #!#!     at java.base/java.util.ArrayList.removeIf(ArrayList.java:1672)
[02:43:38 ERROR]: #!#!     at java.base/java.util.ArrayList.removeIf(ArrayList.java:1660)
[02:43:38 ERROR]: #!#!     at Skript.jar//ch.njol.skript.ScriptLoader.lambda$loadScripts$10(ScriptLoader.java:553)
[02:43:38 ERROR]: #!#!     at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:684)
[02:43:38 ERROR]: #!#!     at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:662)
[02:43:38 ERROR]: #!#!     at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2168)
[02:43:38 ERROR]: #!#!     at Skript.jar//ch.njol.skript.ScriptLoader.loadScripts(ScriptLoader.java:513)
[02:43:38 ERROR]: #!#!     at Skript.jar//ch.njol.skript.ScriptLoader.loadScripts(ScriptLoader.java:459)
[02:43:38 ERROR]: #!#!     at Skript.jar//ch.njol.skript.Skript$1.lambda$run$3(Skript.java:666)
[02:43:38 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_19_R2.scheduler.CraftTask.run(CraftTask.java:101)
[02:43:38 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_19_R2.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:483)
[02:43:38 ERROR]: #!#!     at net.minecraft.server.MinecraftServer.b(MinecraftServer.java:1473)
[02:43:38 ERROR]: #!#!     at net.minecraft.server.dedicated.DedicatedServer.b(DedicatedServer.java:447)
[02:43:38 ERROR]: #!#!     at net.minecraft.server.MinecraftServer.a(MinecraftServer.java:1397)
[02:43:38 ERROR]: #!#!     at net.minecraft.server.MinecraftServer.w(MinecraftServer.java:1173)
[02:43:38 ERROR]: #!#!     at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:316)
[02:43:38 ERROR]: #!#!     at java.base/java.lang.Thread.run(Thread.java:833)
```

Reply:
it makes sense to have because then if a similar mistake is made in the future it'll be identified faster and not silently fail.

Reply:
The enum exists in the same class. It's not like the enum is external. If you're updating the enum in the same class, the same switch cases should be assumed to be easily implemented. This is why the warning "missing cases" appears, but it won't with this weird IllegalStateException.